### DISCUSSION TITLE
[proxyDHCP] [UEFI] What is "pxebs" ? How can I disable it?

### DISCUSSION THREAD

**Discussion by ygxxii** [ipxe/ipxe#750] [Q&A]
- iPXE version: `ipxe.efi` 1.21.1+ (g64917) from official
- dnsmasq version: 2.86

proxyDHCP Server + UEFI PXE Client. and I decided not using `autoexec.ipxe`.

I tried to config multiple `pxe-service` in dnsmasq to specify multiple PXE boot menu items:

```conf
# enable proxyDHCP:
dhcp-range=<IP Range>,proxy

# iPXE sends a 175 option, checking if exists
dhcp-match=set:ipxe,175

pxe-service=tag:!ipxe,x86-64_EFI,"ipxe.efi official",ipxe/x64-ori-ipxe.efi,<TFTP Server IP>
pxe-service=tag:!ipxe,x86-64_EFI,"ipxe.efi customize",ipxe/x64-cus-ipxe.efi,<TFTP Server IP>
pxe-service=tag:!ipxe,x86-64_EFI,"Exit menu"

dhcp-boot=tag:ipxe,http://<HTTP Server IP>/repos/shell.ipxe,,0.0.0.1
```

UEFI PXE boot to iPXE successfully, **but iPXE boot to `tftp://<TFTP Server IP>/ipxe/x64-ori-ipxe.efi` ,not `http://<HTTP Server IP>/repos/shell.ipxe`.**

---

So I tired a test with the same config:

1. Reboot and press `Ctrl+B` enter iPXE Shell, and found there is a interface named `pxebs`:

```
iPXE> config
```

- `net0.dhcp`:
    - `gateway`: `<DHCP Server>`
    - `ip`: `<IP>`
    - `netmask`: `255.255.255.0`
    - `dns`: `<DNS Server>`
    - `next-server`: `<DHCP Server>`
    - `dhcp-server`: `<DHCP Server>`
- `proxydhcp`:
    - `next-server`: `<TFTP Server>`
    - `vender-class`: `PXEClient`
    - `dhcp-server`: `<TFTP Server>`
- `pxebs`:
    - `ip`: `<IP>`
    - `filename`: `ipxe/x64-ori-ipxe.efi`
    - `next-server`: `<TFTP Server>`
    - `vender-class`: `PXEClient`
    - `dhcp-server`: `<TFTP Server>`

2. Don't reboot, run `dhcp`:

```
iPXE> dhcp
iPXE> config
```

- `pxebs`: (unchanged)
    - `ip`: `<IP>`
    - `filename`: `ipxe/x64-ori-ipxe.efi`
    - `next-server`: `<TFTP Server>`
    - `vender-class`: `PXEClient`
    - `dhcp-server`: `<TFTP Server>`
- `net0.dhcp`: (unchanged)
    - `gateway`: `<DHCP Server>`
    - `ip`: `<IP>`
    - `netmask`: `255.255.255.0`
    - `dns`: `<DNS Server>`
    - `next-server`: `<DHCP Server>`
    - `dhcp-server`: `<DHCP Server>`
- `proxydhcp`:
    - `filename`: `http://<HTTP Server IP>/repos/shell.ipxe` (changed)
    - `next-server`: `0.0.0.1` (changed)
    - `vender-class`: `PXEClient`
    - `dhcp-server`: `<TFTP Server>`

---

Then I tried another test: config only one `pxe-service` in dnsmasq:

```conf
# enable proxyDHCP:
dhcp-range=<IP Range>,proxy

# iPXE sends a 175 option, checking if exists
dhcp-match=set:ipxe,175

pxe-service=tag:!ipxe,x86-64_EFI,"ipxe.efi official",ipxe/x64-ori-ipxe.efi,<TFTP Server IP>
#pxe-service=tag:!ipxe,x86-64_EFI,"ipxe.efi customize",ipxe/x64-cus-ipxe.efi,<TFTP Server IP>
#pxe-service=tag:!ipxe,x86-64_EFI,"Exit menu"

dhcp-boot=tag:ipxe,http://<HTTP Server IP>/repos/shell.ipxe,,0.0.0.1
```

Everything was fine: UEFI PXE boot to iPXE, iPXE boot to `http://<HTTP Server IP>/repos/shell.ipxe`.

No `pxebs` exists:

- `net0.dhcp`:
    - `gateway`: `<DHCP Server>`
    - `ip`: `<IP>`
    - `netmask`: `255.255.255.0`
    - `dns`: `<DNS Server>`
    - `next-server`: `<DHCP Server>`
    - `dhcp-server`: `<DHCP Server>`
- `proxydhcp`:
    - `filename`: `http://<HTTP Server IP>/repos/shell.ipxe`
    - `next-server`: `0.0.0.1`
    - `vender-class`: `PXEClient`
    - `dhcp-server`: `<TFTP Server>`

---

Question:

1. What is `pxebs`?
2. How can I disable `pxebs`? I tried build with `#define PXEBS_START_TIMEOUT_SEC 0 ` `#define PXEBS_END_TIMEOUT_SEC 0` `#define PXEBS_MAX_TIMEOUT_SEC 0` in `dhcp.h` but failed.

**Comment by NiKiZe** [comment]
PXEBS or PXE Boot Service is a menu system which mainly was used before more advanced menu systems existed.

In dnsmasq you use pxe-service which provides this information.
iPXE does not support this in EFI, but there is limited support for it in pcbios.

**Reply by stappersg** [reply]
> iPXE does not support this in EFI, but there is limited support for it in pcbios.

That in my words:
> In times of EFI being the _"standard"_, it doesn't make sense to spend energy on `pxebs` and such stuff from a previous era. And if you fancy `pxebs`, use it a PCBIOS environment.


(This update was written after seeing in #909 a reference to this discussion)
